# LIVING IN LONDON: For the everyday person

## Capstone Project - The Battle of Neighborhoods (Week 2)

### For the IBM Applied Data Science Capstone

## 1. Import libraries

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
!pip install geocoder
import geocoder

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


## 2. Download and Explore Initial Dataset

### 2.1 Downloading Dataset
Here is the link to the dataset: https://en.wikipedia.org/wiki/List_of_areas_of_London

In [42]:
#Using Pandas to extract data
data = pd.read_html('https://en.wikipedia.org/wiki/List_of_areas_of_London')

In [43]:
data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
                                            Location  \
 0                                        Abbey Wood   
 1                                             Acton   
 2                                         Addington   
 3                                        Addiscombe   
 4                                       Albany Park   
 5                                  Aldborough Hatch   
 6                                           Aldgate   
 7                                           Aldwych   
 8                                          Alperton   
 9                                           Anerley   
 10                                            Angel   
 11                                        Aperfield   
 12                                          Archway   
 13                                   Ardleigh Green 

In [44]:
df=data[1]
df.head()

Location                     London borough       Post town  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2    Addington                         Croydon[8]         CROYDON   
3   Addiscombe                         Croydon[8]         CROYDON   
4  Albany Park                             Bexley  BEXLEY, SIDCUP   

  Postcode district Dial code OS grid ref  
0               SE2       020    TQ465785  
1            W3, W4       020    TQ205805  
2               CR0       020    TQ375645  
3               CR0       020    TQ345665  
4         DA5, DA14       020    TQ478728

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 533 entries, 0 to 532
Data columns (total 6 columns):
Location             533 non-null object
London borough       533 non-null object
Post town            533 non-null object
Postcode district    533 non-null object
Dial code            533 non-null object
OS grid ref          531 non-null object
dtypes: object(6)
memory usage: 25.1+ KB


### 2.2 Data Cleaning

In [46]:
#Checking location of null values
df.loc[df['OS grid ref'].isnull()]

Location London borough  Post town Postcode district Dial code  \
53     Blendon         Bexley     BEXLEY              DA 5       020   
233  Hazelwood        Bromley  ORPINGTON               BR6     01689   

    OS grid ref  
53          NaN  
233         NaN

In [47]:
#Removing rows in 'OS grid ref' with 'NaN'
df.dropna(subset=['OS grid ref'], axis=0, inplace=True)
df['OS grid ref'].isnull().any()

False

In [48]:
#Checking column names
print(df.columns.tolist())

['Location', 'London\xa0borough', 'Post town', 'Postcode\xa0district', 'Dial\xa0code', 'OS grid ref']


In [49]:
#Keeping the Location, Borough and Postcode variables
df=df[['Location', 'London\xa0borough', 'Postcode\xa0district']]
df.head()

Location                     London borough Postcode district
0   Abbey Wood              Bexley, Greenwich [7]               SE2
1        Acton  Ealing, Hammersmith and Fulham[8]            W3, W4
2    Addington                         Croydon[8]               CR0
3   Addiscombe                         Croydon[8]               CR0
4  Albany Park                             Bexley         DA5, DA14

In [50]:
#Renaming columnns
df.rename(columns={'Location': 'Neighborhood', 'London\xa0borough': 'Borough', 'Postcode\xa0district': 'PostalCode'}, inplace=True)

In [51]:
#Removing all parentheses, brackets and texts in between
df['Neighborhood'] = df['Neighborhood'].str.replace(r"\(.*\)", "")
df['Borough'] = df['Borough'].str.replace(r"\[.*\]", "")
df['Neighborhood'] = df['Neighborhood'].str.strip()
df.head()

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See th

Neighborhood                         Borough PostalCode
0   Abbey Wood              Bexley, Greenwich         SE2
1        Acton  Ealing, Hammersmith and Fulham     W3, W4
2    Addington                         Croydon        CR0
3   Addiscombe                         Croydon        CR0
4  Albany Park                          Bexley  DA5, DA14

In [52]:
#Checking Bexley row which was initially "Bexley (also Old Bexley, Bexley Village)"
df.loc[df['Neighborhood']=='Bexley']

Neighborhood Borough PostalCode
44       Bexley  Bexley        DA5

In [53]:
#Resetting index
df.reset_index(drop=True, inplace=True)
df.loc[53,:]

Neighborhood    Bloomsbury
Borough             Camden
PostalCode             WC1
Name: 53, dtype: object

In [54]:
df.loc[52:54,:]

Neighborhood        Borough PostalCode
52    Blackwall  Tower Hamlets        E14
53   Bloomsbury         Camden        WC1
54   Botany Bay        Enfield        EN2

#### Confirms that Blendon was taken out and dataset is clean.

## 3. Get the Geographical coordinates

In [55]:
#Define a function to get coordinates
def get_latlng(location):
    #Initialize your variable to None
    lat_lng_coords = None
    #Loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, London, United Kingdom'.format(location))
        lat_lng_coords = g.latlng
    return lat_lng_coords

#Call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(location) for location in df["Neighborhood"].tolist() ]

In [56]:
coords[0:5]

[[51.492450000000076, 0.12127000000003818],
 [51.51324000000005, -0.2674599999999714],
 [51.575810032233505, -0.10933991526687237],
 [51.472748982987284, -0.2033256827571753],
 [51.48582000000005, -0.08025999999995292]]

In [57]:
#Create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [58]:
#Merge the coordinates into the original dataframe
df['Latitude'] = df_coords['Latitude']
df['Longitude'] = df_coords['Longitude']

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


In [59]:
df.head()

Neighborhood                         Borough PostalCode   Latitude  \
0   Abbey Wood              Bexley, Greenwich         SE2  51.492450   
1        Acton  Ealing, Hammersmith and Fulham     W3, W4  51.513240   
2    Addington                         Croydon        CR0  51.575810   
3   Addiscombe                         Croydon        CR0  51.472749   
4  Albany Park                          Bexley  DA5, DA14  51.485820   

   Longitude  
0   0.121270  
1  -0.267460  
2  -0.109340  
3  -0.203326  
4  -0.080260

In [60]:
df.shape

(531, 5)

In [61]:
# save the DataFrame as CSV file
df.to_csv("london_df.csv", index=False)

## 4. Explore and Cluster Neighborhoods

### 4.1 Explore locations in London through visualisations
Use geopy to get the latitude and longitude of London, UK.

In [62]:
address = 'London, UK'

geolocator = Nominatim(user_agent="London")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London are 51.5073219, -0.1276474.


Create a map of London with neighborhoods superimposed on top.

In [63]:
map_london = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, location in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(location, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  

map_london

### 4.2 Explore the locations with Foursquare API
Define Foursquare Credentials and Version

In [64]:
CLIENT_ID = '5RH1MJBVVEE12ZPDE4XE0JTS1POSL2WLCMW0LVO5NNASSIX1' # your Foursquare ID
CLIENT_SECRET = 'NYPXTZBGBK2GFKGLOR4LOTXGBY2S33XCBU2ZBYLVPCBJ5MTW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5RH1MJBVVEE12ZPDE4XE0JTS1POSL2WLCMW0LVO5NNASSIX1
CLIENT_SECRET:NYPXTZBGBK2GFKGLOR4LOTXGBY2S33XCBU2ZBYLVPCBJ5MTW


Define Get Nearby Venues function

In [65]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [66]:
London_venues = getNearbyVenues(names=df['Neighborhood'],
                                 latitudes=df['Latitude'],
                                 longitudes=df['Longitude'])

Abbey Wood
Acton
Addington
Addiscombe
Albany Park
Aldborough Hatch
Aldgate
Aldwych
Alperton
Anerley
Angel
Aperfield
Archway
Ardleigh Green
Arkley
Arnos Grove
Balham
Bankside
Barbican
Barking
Barkingside
Barnehurst
Barnes
Barnes Cray
Barnet Gate
Barnet
Barnsbury
Battersea
Bayswater
Beckenham
Beckton
Becontree
Becontree Heath
Beddington
Bedford Park
Belgravia
Bellingham
Belmont
Belmont
Belsize Park
Belvedere
Bermondsey
Berrylands
Bethnal Green
Bexley
Bexleyheath
Bickley
Biggin Hill
Blackfen
Blackfriars
Blackheath
Blackheath Royal Standard
Blackwall
Bloomsbury
Botany Bay
Bounds Green
Bow
Bowes Park
Brentford
Brent Cross
Brent Park
Brimsdown
Brixton
Brockley
Bromley
Bromley
Bromley Common
Brompton
Brondesbury
Brunswick Park
Bulls Cross
Burnt Oak
Burroughs, The
Camberwell
Cambridge Heath
Camden Town
Canary Wharf
Cann Hall
Canning Town
Canonbury
Carshalton
Castelnau
Castle Green
Catford
Chadwell Heath
Chalk Farm
Charing Cross
Charlton
Chase Cross
Cheam
Chelsea
Chelsfield
Chessington
Childs H

Checkign number of venue returned for each neighborhood

In [67]:
London_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                               
Abbey Wood                                     7                       7   
Acton                                          7                       7   
Addington                                      7                       7   
Addiscombe                                    63                      63   
Albany Park                                    8                       8   
Aldborough Hatch                               4                       4   
Aldgate                                       71                      71   
Aldwych                                       80                      80   
Alperton                                      18                      18   
Anerley                                        5                       5   
Angel                                        100                     100   
Aperfield                                     12                      12   
Archway                                       35                      35   
Ardleigh Green                                36                      36   
Arkley                                        15                      15   
Arnos Grove                                    7                       7   
Balham                                        36                      36   
Bankside                                     100                     100   
Barbican                                      42                      42   
Barking                                        4                       4   
Barkingside                                    4                       4   
Barnehurst                                     4                       4   
Barnes                                        23                      23   
Barnes Cray                                   23                      23   
Barnet                                         1                       1   
Barnet Gate                                    3                       3   
Barnsbury                                     68                      68   
Battersea                                     26                      26   
Bayswater                                     47                      47   
Beckenham                                      4                       4   
Beckton                                       16                      16   
Becontree                                     25                      25   
Becontree Heath                               25                      25   
Beddington                                     4                       4   
Bedford Park                                  14                      14   
Belgravia                                     39                      39   
Bellingham                                     4                       4   
Belmont                                       10                      10   
Belsize Park                                  61                      61   
Belvedere                                     28                      28   
Bermondsey                                    48                      48   
Berrylands                                     4                       4   
Bethnal Green                                 36                      36   
Bexley                                         5                       5   
Bexleyheath                                   16                      16   
Bickley                                       24                      24   
Biggin Hill                                    3                       3   
Blackfen                                      63                      63   
Blackfriars                                   70                      70   
Blackheath                                     2                       2   
Blackheath Royal Standard                      6                       6   

Encoding Categories of Venues

In [68]:
#One hot encoding
London_onehot = pd.get_dummies(London_venues[['Venue Category']], prefix="", prefix_sep="")

#Adding Neighborhood column to dataframe
London_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
London_onehot.insert(loc=0, column='Neighborhood', value=London_venues['Neighborhood'] )

#Getting the shape of the dataframe
London_onehot.shape

(14103, 415)

Group rows by neighborhood and take mean of the frequency of each category

In [69]:
London_grouped = London_onehot.groupby('Neighborhood').mean().reset_index()
London_grouped.head()

Neighborhood  ATM  Accessories Store  Adult Boutique  Afghan Restaurant  \
0   Abbey Wood  0.0                0.0             0.0                0.0   
1        Acton  0.0                0.0             0.0                0.0   
2    Addington  0.0                0.0             0.0                0.0   
3   Addiscombe  0.0                0.0             0.0                0.0   
4  Albany Park  0.0                0.0             0.0                0.0   

   African Restaurant  Airport Service  Airport Terminal  American Restaurant  \
0                 0.0              0.0               0.0                  0.0   
1                 0.0              0.0               0.0                  0.0   
2                 0.0              0.0               0.0                  0.0   
3                 0.0              0.0               0.0                  0.0   
4                 0.0              0.0               0.0                  0.0   

   Antique Shop  Aquarium  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0           0.0       0.0     0.0               0.0                     0.0   
1           0.0       0.0     0.0               0.0                     0.0   
2           0.0       0.0     0.0               0.0                     0.0   
3           0.0       0.0     0.0               0.0                     0.0   
4           0.0       0.0     0.0               0.0                     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0          0.0         0.0                  0.0                   0.0   
1          0.0         0.0                  0.0                   0.0   
2          0.0         0.0                  0.0                   0.0   
3          0.0         0.0                  0.0                   0.0   
4          0.0         0.0                  0.0                   0.0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  \
0               0.0            0.000000                    0.0   
1               0.0            0.000000                    0.0   
2               0.0            0.000000                    0.0   
3               0.0            0.015873                    0.0   
4               0.0            0.000000                    0.0   

   Austrian Restaurant  Auto Dealership  Auto Garage  Auto Workshop  \
0                  0.0              0.0          0.0            0.0   
1                  0.0              0.0          0.0            0.0   
2                  0.0              0.0          0.0            0.0   
3                  0.0              0.0          0.0            0.0   
4                  0.0              0.0          0.0            0.0   

   Automotive Shop  BBQ Joint  Bagel Shop    Bakery    Bar  Baseball Field  \
0              0.0        0.0         0.0  0.000000  0.000             0.0   
1              0.0        0.0         0.0  0.000000  0.000             0.0   
2              0.0        0.0         0.0  0.000000  0.000             0.0   
3              0.0        0.0         0.0  0.063492  0.000             0.0   
4              0.0        0.0         0.0  0.000000  0.125             0.0   

   Basketball Court  Beach  Bed & Breakfast  Beer Bar  Beer Garden  \
0               0.0    0.0         0.000000       0.0          0.0   
1               0.0    0.0         0.142857       0.0          0.0   
2               0.0    0.0         0.000000       0.0          0.0   
3               0.0    0.0         0.000000       0.0          0.0   
4               0.0    0.0         0.000000       0.0          0.0   

   Beer Store  Betting Shop  Bike Rental / Bike Share  Bike Shop  Bistro  \
0         0.0           0.0                       0.0        0.0     0.0   
1         0.0           0.0                       0.0        0.0     0.0   
2         0.0           0.0                       0.0        0.0     0.0   
3         0.0           0.0                       0.0        0.0     0.0   
4         0.0           0.0                 

## 5. Analysis for Supermarket and Train Station

Create new dataframe with supermarket and train station and conduct clustering.

In [70]:
#smts stands for supermarket and train station
London_smts=London_grouped[['Neighborhood', 'Supermarket', 'Train Station']]

Making clusters for neighborhoods

In [71]:
#Set number of clusters
kclusters = 5

London_smts_clustering = London_smts.drop('Neighborhood', 1)

#Run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(London_smts_clustering)

#Check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 2, 2, 1, 1, 1, 1, 1, 1, 4], dtype=int32)

In [72]:
#Create new dataset for merging
London_merged = London_smts

#Add clustering labels
London_merged['Clusters']=kmeans.labels_
London_merged.head()

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Neighborhood  Supermarket  Train Station  Clusters
0   Abbey Wood     0.285714       0.142857         3
1        Acton     0.000000       0.142857         2
2    Addington     0.000000       0.142857         2
3   Addiscombe     0.015873       0.000000         1
4  Albany Park     0.000000       0.000000         1

Adding most common venues in locations.
Define function to sort venues in descending order

In [73]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create a dataframe that displays the top 10 venues for each neighborhood

In [74]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = London_grouped['Neighborhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0   Abbey Wood           Supermarket         Train Station   
1        Acton         Grocery Store        Breakfast Spot   
2    Addington                  Café                 Trail   
3   Addiscombe    Italian Restaurant                  Café   
4  Albany Park                  Café              Building   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop     Convenience Store              Platform   
1     Indian Restaurant       Bed & Breakfast                  Park   
2                  Park           Coffee Shop      Tapas Restaurant   
3           Coffee Shop                Bakery                   Pub   
4                   Pub  Fast Food Restaurant                Garden   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Historic Site  Ethiopian Restaurant         Event Service   
1         Train Station           Zoo Exhibit   Filipino Restaurant   
2         Train Station     Convenience Store            Food Court   
3         Grocery Store             Juice Bar          Climbing Gym   
4                   Bar   Sporting Goods Shop          Food Service   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space                Exhibit  
1               Exhibit            Fabric Shop  
2  Fast Food Restaurant          Event Service  
3                  Park        Thai Restaurant  
4           Film Studio                Exhibit

Merging Cluster dataset and Most Common Venue dataset

In [75]:
London= London_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
London= London.join(df.set_index('Neighborhood'), on='Neighborhood')
London.head()

Neighborhood  Supermarket  Train Station  Clusters 1st Most Common Venue  \
0   Abbey Wood     0.285714       0.142857         3           Supermarket   
1        Acton     0.000000       0.142857         2         Grocery Store   
2    Addington     0.000000       0.142857         2                  Café   
3   Addiscombe     0.015873       0.000000         1    Italian Restaurant   
4  Albany Park     0.000000       0.000000         1                  Café   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Train Station           Coffee Shop     Convenience Store   
1        Breakfast Spot     Indian Restaurant       Bed & Breakfast   
2                 Trail                  Park           Coffee Shop   
3                  Café           Coffee Shop                Bakery   
4              Building                   Pub  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Platform         Historic Site  Ethiopian Restaurant   
1                  Park         Train Station           Zoo Exhibit   
2      Tapas Restaurant         Train Station     Convenience Store   
3                   Pub         Grocery Store             Juice Bar   
4                Garden                   Bar   Sporting Goods Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0         Event Service           Event Space                Exhibit   
1   Filipino Restaurant               Exhibit            Fabric Shop   
2            Food Court  Fast Food Restaurant          Event Service   
3          Climbing Gym                  Park        Thai Restaurant   
4          Food Service           Film Studio                Exhibit   

                          Borough PostalCode   Latitude  Longitude  
0              Bexley, Greenwich         SE2  51.492450   0.121270  
1  Ealing, Hammersmith and Fulham     W3, W4  51.513240  -0.267460  
2                         Croydon        CR0  51.575810  -0.109340  
3                         Croydon        CR0  51.472749  -0.203326  
4                          Bexley  DA5, DA14  51.485820  -0.080260

## 6. Results

### 6.1 Creating a new map

In [76]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(London['Latitude'], London['Longitude'], London['Neighborhood'], London['Clusters']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 6.2 Viewing individual clusters

In [77]:
#Cluster 0
London.loc[London['Clusters'] == 0]

Neighborhood  Supermarket  Train Station  Clusters  \
14                 Arkley     0.133333       0.000000         0   
27              Battersea     0.076923       0.038462         0   
30                Beckton     0.062500       0.000000         0   
44            Bexleyheath     0.062500       0.000000         0   
45                Bickley     0.083333       0.000000         0   
57            Brent Cross     0.066667       0.000000         0   
58             Brent Park     0.142857       0.000000         0   
85               Charlton     0.133333       0.000000         0   
101             Colindale     0.142857       0.000000         0   
105               Colyers     0.111111       0.000000         0   
107                Coombe     0.058824       0.000000         0   
107                Coombe     0.058824       0.000000         0   
123           Cubitt Town     0.058824       0.000000         0   
124                Cudham     0.117647       0.000000         0   
144              East Ham     0.071429       0.000000         0   
147              Eastcote     0.080000       0.000000         0   
166              Finchley     0.100000       0.000000         0   
177               Frognal     0.058824       0.000000         0   
187          Grahame Park     0.125000       0.000000         0   
234              Holloway     0.057143       0.000000         0   
235              Homerton     0.085714       0.028571         0   
239            Hornchurch     0.166667       0.000000         0   
245             Isleworth     0.066667       0.000000         0   
247                Kenley     0.166667       0.000000         0   
256               Kilburn     0.064516       0.032258         0   
258             Kingsbury     0.090909       0.000000         0   
273           Leytonstone     0.083333       0.000000         0   
289              Maryland     0.083333       0.000000         0   
303              Mortlake     0.052632       0.000000         0   
317             Nine Elms     0.083333       0.000000         0   
320               Norbury     0.090909       0.000000         0   
323        North Finchley     0.062500       0.000000         0   
330  Northumberland Heath     0.166667       0.000000         0   
346            Park Royal     0.111111       0.000000         0   
377               Ruislip     0.069767       0.000000         0   
379                Ruxley     0.100000       0.000000         0   
398            Somerstown     0.076923       0.000000         0   
402      South Hornchurch     0.166667       0.000000         0   
404         South Norwood     0.055556       0.000000         0   
425         Stamford Hill     0.090909       0.000000         0   
435               Sudbury     0.111111       0.000000         0   
441              Sydenham     0.076923       0.000000         0   
445        Temple Fortune     0.090909       0.000000         0   
448              The Hyde     0.083333       0.000000         0   
452               Tooting     0.056604       0.000000         0   
454             Tottenham     0.125000       0.000000         0   
468          Upper Ruxley     0.100000       0.000000         0   
471            Upton Park     0.076923       0.000000         0   
473              Vauxhall     0.058824       0.000000         0   
479            Wandsworth     0.058824       0.000000         0   
484               Welling     0.176471       0.000000         0   
496            West Heath     0.083333       0.000000         0   
499          West Norwood     0.050000       0.000000         0   
510        Winchmore Hill     0.142857       0.000000         0   
517              Woolwich     0.069767       0.000000         0   
518        Worcester Park     0.125000       0.000000         0   
521              Yiewsley     0.100000       0.050000         0   

       1st Most Common Venue   2nd Most Common Venue  \
14               Supermarket           Grocery Store   
27            

In [78]:
#Cluster 1
London.loc[London['Clusters'] == 1]

Neighborhood  Supermarket  Train Station  Clusters  \
3                   Addiscombe     0.015873       0.000000         1   
4                  Albany Park     0.000000       0.000000         1   
5             Aldborough Hatch     0.000000       0.000000         1   
6                      Aldgate     0.000000       0.000000         1   
7                      Aldwych     0.000000       0.000000         1   
8                     Alperton     0.000000       0.000000         1   
10                       Angel     0.010000       0.000000         1   
11                   Aperfield     0.000000       0.000000         1   
12                     Archway     0.000000       0.000000         1   
13              Ardleigh Green     0.000000       0.000000         1   
15                 Arnos Grove     0.000000       0.000000         1   
16                      Balham     0.027778       0.000000         1   
17                    Bankside     0.000000       0.000000         1   
18                    Barbican     0.023810       0.000000         1   
19                     Barking     0.000000       0.000000         1   
21                  Barnehurst     0.000000       0.000000         1   
22                      Barnes     0.000000       0.000000         1   
23                 Barnes Cray     0.000000       0.000000         1   
24                      Barnet     0.000000       0.000000         1   
25                 Barnet Gate     0.000000       0.000000         1   
26                   Barnsbury     0.029412       0.000000         1   
28                   Bayswater     0.000000       0.000000         1   
29                   Beckenham     0.000000       0.000000         1   
31                   Becontree     0.040000       0.000000         1   
32             Becontree Heath     0.040000       0.000000         1   
33                  Beddington     0.000000       0.000000         1   
34                Bedford Park     0.000000       0.000000         1   
35                   Belgravia     0.025641       0.000000         1   
37                     Belmont     0.000000       0.000000         1   
37                     Belmont     0.000000       0.000000         1   
38                Belsize Park     0.016393       0.000000         1   
39                   Belvedere     0.000000       0.000000         1   
40                  Bermondsey     0.000000       0.000000         1   
41                  Berrylands     0.000000       0.000000         1   
42               Bethnal Green     0.000000       0.000000         1   
43                      Bexley     0.000000       0.000000         1   
46                 Biggin Hill     0.000000       0.000000         1   
47                    Blackfen     0.000000       0.000000         1   
48                 Blackfriars     0.000000       0.000000         1   
49                  Blackheath     0.000000       0.000000         1   
50   Blackheath Royal Standard     0.000000       0.000000         1   
51                   Blackwall     0.000000       0.000000         1   
52                  Bloomsbury     0.021739       0.000000         1   
53                  Botany Bay     0.000000       0.000000         1   
54                Bounds Green     0.000000       0.000000         1   
55                         Bow     0.000000       0.000000         1   
56                  Bowes Park     0.000000       0.000000         1   
59                   Brentford     0.000000       0.000000         1   
60                   Brimsdown     0.000000       0.000000         1   
61                     Brixton     0.000000       0.000000         1   
62                    Brockley     0.000000       0.028571         1   
65                    Brompton     0.000000       0.000000         1   
66                 Brondesbury     0.000000       0.000000         1   
68                 Bulls Cross     0.000000       0.000000         1   
69                   Burnt Oak     0.000000       0.000000

In [79]:
#Cluster 3
London.loc[London['Clusters'] == 2]

Neighborhood  Supermarket  Train Station  Clusters  \
1                Acton     0.000000       0.142857         2   
2            Addington     0.000000       0.142857         2   
63             Bromley     0.076923       0.076923         2   
63             Bromley     0.076923       0.076923         2   
64      Bromley Common     0.076923       0.076923         2   
67      Brunswick Park     0.000000       0.066667         2   
77           Canonbury     0.000000       0.055556         2   
93           Chinbrook     0.000000       0.076923         2   
111           Cranford     0.000000       0.142857         2   
134              Downe     0.000000       0.142857         2   
136            Dulwich     0.000000       0.125000         2   
150           Edmonton     0.000000       0.090909         2   
154         Elmers End     0.111111       0.111111         2   
163         Falconwood     0.000000       0.142857         2   
172        Forest Hill     0.000000       0.055556         2   
186         Gospel Oak     0.000000       0.062500         2   
207            Hanwell     0.071429       0.071429         2   
210          Harlesden     0.000000       0.166667         2   
213        Harold Hill     0.000000       0.052632         2   
214        Harold Park     0.000000       0.052632         2   
216          Harringay     0.000000       0.083333         2   
231       Hither Green     0.000000       0.083333         2   
252             Kenton     0.058824       0.058824         2   
265            Lampton     0.000000       0.090909         2   
266         Lea Bridge     0.125000       0.125000         2   
304       Motspur Park     0.000000       0.111111         2   
309      New Addington     0.000000       0.142857         2   
373        Riddlesdown     0.000000       0.125000         2   
378         Rush Green     0.000000       0.076923         2   
380        Sanderstead     0.090909       0.090909         2   
382           Selhurst     0.000000       0.125000         2   
385      Seven Sisters     0.000000       0.058824         2   
430        Stonebridge     0.000000       0.066667         2   
432    Strawberry Hill     0.000000       0.083333         2   
451           Tolworth     0.000000       0.055556         2   
469  Upper Walthamstow     0.000000       0.083333         2   
470              Upton     0.000000       0.066667         2   
501     Westcombe Park     0.000000       0.090909         2   

    1st Most Common Venue      2nd Most Common Venue  \
1           Grocery Store             Breakfast Spot   
2                    Café                      Trail   
63    Sporting Goods Shop             Soccer Stadium   
63    Sporting Goods Shop             Soccer Stadium   
64    Sporting Goods Shop             Soccer Stadium   
67            Coffee Shop                        Pub   
77                    Pub                      Trail   
93               Platform              Grocery Store   
111      Tapas Restaurant                Coffee Shop   
134         Train Station                     Bistro   
136         Train Station                  Wine Shop   
150         Train Station                 Soup Place   
154         Grocery Store              Train Station   
163     Other Repair Shop                      Trail   
172                   Pub                       Café   
186                  Café         Italian Restaurant   
207                  Café                        Pub   
210   Rental Car Location         Chinese Restaurant   
213           Coffee Shop                        Gym   
214           Coffee Shop                        Gym   
216       Nature Preserve                       Park   
231                  Café         Italian Restaurant   
252                  Café                        Pub   
265                   Pub         Italian Restaurant   
266         Train Station                Coffee Shop   
304                  Park              Train Station   
3

In [80]:
#Cluster 4
London.loc[London['Clusters'] == 3]

Neighborhood  Supermarket  Train Station  Clusters  \
0       Abbey Wood     0.285714       0.142857         3   
20     Barkingside     0.500000       0.000000         3   
86     Chase Cross     0.333333       0.000000         3   
116    Cricklewood     0.250000       0.000000         3   
118      Crook Log     0.200000       0.000000         3   
121        Croydon     0.250000       0.000000         3   
126       Dagenham     0.200000       0.000000         3   
149        Edgware     0.200000       0.000000         3   
218   Harrow Weald     0.200000       0.000000         3   
329       Northolt     0.250000       0.000000         3   
349          Penge     0.333333       0.000000         3   
358    Ponders End     0.333333       0.000000         3   
399  South Croydon     0.250000       0.000000         3   
475     Wallington     0.250000       0.250000         3   
500   West Wickham     0.250000       0.000000         3   

    1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0             Supermarket         Train Station            Coffee Shop   
20            Supermarket         Grocery Store                   Park   
86                   Park           Supermarket            Zoo Exhibit   
116           Supermarket           Bus Station            Pizza Place   
118         Grocery Store           Supermarket  Australian Restaurant   
121                  Café           Supermarket                   Park   
126           Auto Garage     Indian Restaurant            Supermarket   
149           Supermarket                   Pub         Sandwich Place   
218         Grocery Store     Indian Restaurant            Supermarket   
329                  Café           Supermarket          Grocery Store   
349           Supermarket                 Hotel      Convenience Store   
358    Turkish Restaurant           Supermarket            Coffee Shop   
399                  Café           Supermarket                   Park   
475            Restaurant                 River            Supermarket   
500           Supermarket           Pizza Place                 Bakery   

    4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0       Convenience Store               Platform         Historic Site   
20            Zoo Exhibit            Film Studio               Exhibit   
86            Film Studio            Event Space               Exhibit   
116                  Park            Zoo Exhibit   Filipino Restaurant   
118                  Park  Vietnamese Restaurant           Zoo Exhibit   
121  Fast Food Restaurant            Zoo Exhibit               Exhibit   
126                   Pub            Gas Station           Zoo Exhibit   
149           Auto Garage   Outdoor Supply Store    Dim Sum Restaurant   
218       Thai Restaurant   Fast Food Restaurant           Zoo Exhibit   
329  Fast Food Restaurant            Zoo Exhibit               Exhibit   
349         Grocery Store   Fast Food Restaurant           Film Studio   
358                   Pub            Zoo Exhibit         Event Service   
399  Fast Food Restaurant            Zoo Exhibit               Exhibit   
475         Train Station            Zoo Exhibit   Filipino Restaurant   
500              Pharmacy      Indian Restaurant                   Pub   

      7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0      Ethiopian Restaurant         Event Service           Event Space   
20              Fabric Shop               Factory    Falafel Restaurant   
86              Fabric Shop               Factory    Falafel Restaurant   
116             Event Space               Exhibit           Fabric Shop   
118     Filipino Restaurant           Event Space               Exhibit   
121             Fabric Shop               Factory    Falafel Restaurant   
126             Film Studio           Event Space               Exhibit   
149  Furniture / Home Store  Gym / Fitness Center    Chinese Restaurant   


In [81]:
#Cluster 5
London.loc[London['Clusters'] == 4]

Neighborhood  Supermarket  Train Station  Clusters  \
9           Anerley          0.0       0.200000         4   
36       Bellingham          0.0       0.250000         4   
81          Catford          0.0       0.200000         4   
197     Hadley Wood          0.0       0.333333         4   
249    Kensal Green          0.0       0.200000         4   
302     Morden Park          0.0       0.333333         4   
375         Romford          0.0       0.250000         4   
513  Woodford Green          0.0       0.222222         4   

    1st Most Common Venue   2nd Most Common Venue 3rd Most Common Venue  \
9          Hardware Store                     Pub           Gas Station   
36                Brewery                Bus Stop         Metro Station   
81         Discount Store  Furniture / Home Store    Turkish Restaurant   
197                Tunnel             Golf Course         Train Station   
249                Bakery                     Pub                  Park   
302         Train Station      English Restaurant                  Pool   
375           Bus Station           Grocery Store                Arcade   
513                  Café                     Pub         Train Station   

     4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
9                     Park         Train Station           Zoo Exhibit   
36           Train Station            Food Truck            Food Stand   
81             Pizza Place         Train Station            Food Court   
197            Zoo Exhibit   Filipino Restaurant         Event Service   
249  Portuguese Restaurant         Train Station   Filipino Restaurant   
302                   Park                 Hotel         Event Service   
375          Train Station           Zoo Exhibit           Film Studio   
513             Restaurant    Chinese Restaurant         Grocery Store   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
9    Fast Food Restaurant         Event Service           Event Space   
36                Exhibit           Fabric Shop               Factory   
81           Food Service  Ethiopian Restaurant         Event Service   
197           Event Space               Exhibit           Fabric Shop   
249         Event Service           Event Space               Exhibit   
302           Event Space               Exhibit           Fabric Shop   
375           Event Space               Exhibit           Fabric Shop   
513          Food Service  Ethiopian Restaurant         Event Service   

    10th Most Common Venue                    Borough PostalCode   Latitude  \
9                  Exhibit                    Bromley       SE20  51.412330   
36                  Forest                   Lewisham        SE6  51.602570   
81             Event Space                   Lewisham        SE6  51.437220   
197                Factory                    Enfield        EN4  51.666690   
249            Fabric Shop                      Brent  NW10, NW6  51.530540   
302                Factory                     Merton        SM4  51.392060   
375                Factory                   Havering        RM1  51.548590   
513            Event Space  Redbridge, Waltham Forest        IG8  51.553463   

     Longitude  
9    -0.065390  
36   -0.055870  
81   -0.018680  
197  -0.169810  
249  -0.225480  
302  -0.203520  
375   0.041170  
513   0.025281

__Result__:

In [82]:
print("Average in Cluster 0 for Supermarkets: %.2f" % London['Supermarket'].loc[London['Clusters'] == 0].mean(axis=0))
print("Average in Cluster 0 for Train Stations: %.2f" % London['Train Station'].loc[London['Clusters'] == 0].mean(axis=0))
print("Number of locations: ", London['Train Station'].loc[London['Clusters'] == 0].size)

Average in Cluster 0 for Supermarkets: 0.09
Average in Cluster 0 for Train Stations: 0.00
Number of locations:  57


In [83]:
print("Average in Cluster 1 for Supermarkets: %.2f" % London['Supermarket'].loc[London['Clusters'] == 1].mean(axis=0))
print("Average in Cluster 1 for Train Stations: %.2f" % London['Train Station'].loc[London['Clusters'] == 1].mean(axis=0))
print("Number of locations: ", London['Train Station'].loc[London['Clusters'] == 1].size)

Average in Cluster 1 for Supermarkets: 0.01
Average in Cluster 1 for Train Stations: 0.00
Number of locations:  411


In [84]:
print("Average in Cluster 2 for Supermarkets: %.2f" % London['Supermarket'].loc[London['Clusters'] == 2].mean(axis=0))
print("Average in Cluster 2 for Train Stations: %.2f" % London['Train Station'].loc[London['Clusters'] == 2].mean(axis=0))
print("Number of locations: ", London['Train Station'].loc[London['Clusters'] == 2].size)

Average in Cluster 2 for Supermarkets: 0.02
Average in Cluster 2 for Train Stations: 0.09
Number of locations:  38


In [85]:
print("Average in Cluster 3 for Supermarkets: %.2f" % London['Supermarket'].loc[London['Clusters'] == 3].mean(axis=0))
print("Average in Cluster 3 for Train Stations: %.2f" % London['Train Station'].loc[London['Clusters'] == 3].mean(axis=0))
print("Number of locations: ", London['Train Station'].loc[London['Clusters'] == 3].size)

Average in Cluster 3 for Supermarkets: 0.27
Average in Cluster 3 for Train Stations: 0.03
Number of locations:  15


In [86]:
print("Average in Cluster 4 for Supermarkets: %.2f" % London['Supermarket'].loc[London['Clusters'] == 4].mean(axis=0))
print("Average in Cluster 4 for Train Stations: %.2f" % London['Train Station'].loc[London['Clusters'] == 4].mean(axis=0))
print("Number of locations: ", London['Train Station'].loc[London['Clusters'] == 4].size)

Average in Cluster 4 for Supermarkets: 0.00
Average in Cluster 4 for Train Stations: 0.25
Number of locations:  8


Results in a table format:

Clusters | Size | Supermarkets(Avg) | Train Stations(Avg)
-|-|-|-
0 | 57  | 0.09 | 0.00
1 | 411 | 0.01 | 0.00
2 | 38  | 0.02 | 0.09
3 | 15  | 0.27 | 0.03
4 | 8   | 0.00 | 0.25